### PR TITLE
Ninjas and Traitors can no longer hack consoles located in maintenance

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- Ninja hacking and Traitor console hacking can only be done in areas with the `VALID_TERRITORY` flag set. This excludes maintenance and solar arrays. 
   - Basically, they will abide by the same rules Space Dragon abide by. 

## Why It's Good For The Game

The hacking objectives are designed at their core to get the antagonist to get involved somewhere deep in the center of the station - they're committing an act of corporate espionage, hacking databases and stuff, it's cool

Printing a board from engineering and making it in the corner of maint is boring, it doesn't encourage the antag to go fight or get involved or sneak around.

At the very least if a traitor wants to build a console to accomplish their objectives, they'll have to build it in a station area, which will likely be in camera range of the AI, allowing them to interject. 

## Changelog

:cl: Melbert
balance: Ninja and Traitors can no longer hack consoles located in maintenance for their objectives. (Basically, it abides by the same rules that Space Dragons have on where they can put their portal.)
fix: Secure Data consoles abide by the same hacking rules as comms consoles - getting de-powered mid-hack will stop hack progress.
/:cl:

### CODE REVIEW

**Comment by Pumpkinoe** [comment]
big fan of making traitors have to come center stage to make this happen, but maybe we could step it up a bit? it would be cool if the check made sure you were in an area of command - since there's still plenty 'out-of-maintenance' places that are just as lame to set up a comms console for example (say, in xenobiology or something)
would provide some fun challenge and espionage imo, though that might be easily out of scope for what this PR wants to accomplish
(just my two cents)

**Comment by thebestcodertoevercode** [comment]
Would love to see this change on the comms console not working in maints as well. Overall good pr.

**Comment by JohnFulpWillard** [comment]
Security nerf wtf

**Comment by Shadowflame909** [comment]
normal: after this change hack into bridge/captains office to complete this objective

gigabrain: put holodeck into star trek mode, destroy one of the holographic consoles and replace it with a comms machine. No one will ever know the difference.

File: code/modules/antagonists/traitor/objectives/hack_comm_console.dm [hunk 1]
```diff
@@ -44,6 +44,10 @@
 	return COMPONENT_CANCEL_ATTACK_CHAIN
 
 /datum/traitor_objective/hack_comm_console/proc/begin_hack(mob/user, obj/machinery/computer/communications/target)
+	if(!target.can_hack())
+		target.balloon_alert(user, "can't hack this console!")
```
Comment (itseasytosee) [inline code/modules/antagonists/traitor/objectives/hack_comm_console.dm:48]: This feedback is quite non discript. There's nothing in game that suggests to the player as to why they can't hack the console, and no obvious in fiction reason.

Comment (itseasytosee) [reply code/modules/antagonists/traitor/objectives/hack_comm_console.dm:48]: something along the lines of Signal too weak, or can't hack in maintenance etc.
Something to make it clear to the players who don't know about this feature WHY not, and bonus points if you can give explanation beyond it being an arbitrary game mechanic.

Comment (Archie700) [reply code/modules/antagonists/traitor/objectives/hack_comm_console.dm:48]: "You can't send the signal from here! The signal is too weak in that area to connect to syndicate channels!"
Something like this.

File: code/game/machinery/computer/communications.dm [hunk 2]
```diff
@@ -776,21 +776,27 @@
...
 
 	hack_console(hacker)
 	return TRUE
 
 /// Checks if this console is hackable. Used as a callback during try_hack_console's doafter as well.
-/obj/machinery/computer/communications/proc/can_hack()
+/obj/machinery/computer/communications/proc/can_hack(mob/living/hacker, feedback = FALSE)
 	if(machine_stat & (NOPOWER|BROKEN))
+		if(feedback && hacker)
+			balloon_alert(hacker, "can't hack this console!")
```
Comment (JohnFulpWillard) [inline code/game/machinery/computer/communications.dm:793]:
```suggestion
			balloon_alert(hacker, "can't hack!")
```

File: code/game/machinery/computer/communications.dm [hunk 3]
```diff
@@ -776,21 +776,27 @@
...
+		return FALSE
+	var/area/console_area = get_area(src)
+	if(!console_area || !(console_area.area_flags & VALID_TERRITORY))
+		if(feedback && hacker)
+			balloon_alert(hacker, "signal too weak, can't hack here!")
```
Comment (JohnFulpWillard) [inline code/game/machinery/computer/communications.dm:798]:
```suggestion
			balloon_alert(hacker, "signal too weak!")
```

Comment (JohnFulpWillard) [reply code/game/machinery/computer/communications.dm:798]: Same for the other console

**Comment by Astrogem2** [comment]
murdering you

**Pull request merged**